README.md
This is an OpenGL 4 remake of the CGA-version of 1987 IBM-Compatible game:

=================================>TEST DRIVE

My Computer Graphics students have to do a `game project' assignment, so I
thought I'd do it too. I'll start a bit ahead of time because I think it's neat
to see /how/ something like this can be made.

I chose Test Drive because it has some interesting elements in 2d and 3d, and
bizarre colouring. I was also thinking of Battlezone for the Atari 2600, but
there's plenty of remakes of that already.

Anton Gerdelan
16 Oct 2014


Changelog
21 dec 2014
* spotted bug in MSAA hint
* reduced end-level text side/position to fit in 800x800

9 Dec 2014
* win32 build, libs

8 Dec 2014 -- version 1.0 -- finished game!
* analogue speedo with needle that works
* gear indicator on dashboard
* finishing screen with final time etc.

18 nov 2014
* upgraded gamepad controls for new motor sim
* tachometer using multi-texturing
* over-revving engine blows it = crash
* fiddled with lambo specs to make it more fun to drive
* sharper steering
* faster trucks
* changed controls for keyboard to cursor keys and q,a, to match original
* improved fov rate of change for max 300kph
* upgraded startup and all shaders to GLSL 4.2.0 and GL 4.2 for better texture
binding

17 nove 2014
* fixed bug - msaa called in wrong place was causing grainy crap on poly edges
* added lamo countach stats for gears/engine etc.
* created update_motor() drive-train simulation
* added kph and rpm and gear text display (temporary)

15 nov 2014
* collision detection with trucks (check all based on distance)
* collision detection with terrain (gets start and end pos of current segment,
and interpolates based on player pos along this. works out wall and cliff drop
x as x = start_pos + gradient * factor +- 1.0 for each side. then checks vs
player x)
* finish level check (is player z < -800)
* fullscreen mode and -res command line params
* hidden mouse cursor
* crash = stop player for 3s and then reposition on road
also immune for 2s after that to prevent insta-death
* smashed windscreen when crashed
* bigger dash
* check if truck/terrain is behind camera if so dont draw
* check if truck/terrain is too far ahead if so dont draw

13 Nov 2014
* added gamepad/joystick detection and controls for xbox 360 controller

10 Nov 2014
* added text rendering for level timer

2 Nov 2014
* steering wheel
* moved mirror along a bit
* added left and right curves randomly to road sections
* fixed dash camera so doesnt actually move with car

31 Oct 2014
* added sounds with irrklang (only linux64 atm)
* pitch shifting of sounds with speed
* change focal length of perspective matrix with speed
* failed attempt at post-processing motion blur

30 Oct 2014
* truck model and truck loader/updater/drawer code
* moving wheels for the trucks (trucks don't drive yet though)
* rear-vision mirror
* moving truck traffic!
* emission-mapped truck headlights

23 Oct 2014 - anton
* entire level's worth of road lengths are generated in a loop and all are
drawn at once (no optimisation yet)

22 Oct 2014 - anton
* added callback with GLFW to handle updating perspective matrix when window is
resized
* added player.c/.h to handle keyboard input etc.
* basic driving controls
* fixed terrain gen so that it's -1 to 1 instead of 0 to 1 on x

16 Oct 2014 - anton
* first commit, project only a plan at this stage
* added slight randomness to X for each segment so that there are no
completely straight road bits - for interesting to look at
* add cliffs to an additional VAO. they join to the side of the road, but i
render separately b/c want a different shader
* cliffs purple shader with screen-space cross-hatching reproduced. scaled-up
for higher-res display. gl_FragCoord.xy to get on-screen position of fragments,
then even/odd to determine if black dot or purple
* added distance fade-to-dark for shaders using gl_FragCoord.w
* added gamma correction to shaders


src/terrain.c
#include "terrain.h"
#include "gl_utils.h"
#include "camera.h"
#include "traffic.h"
#include "player.h"
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

#define ROAD_VS "shaders/road.vert"
#define ROAD_FS "shaders/road.frag"
#define ROAD_TEX "textures/road_seg.png"
#define RCLIFF_VS "shaders/rcliff.vert"
#define RCLIFF_FS "shaders/rcliff.frag"
#define ROADS_PER_LEVEL 20
#define MAX_TERRAIN_DRAW_DIST 20.0

// VAOs pointing to generated mesh data
GLuint road_vaos[ROADS_PER_LEVEL], rcliff_vaos[ROADS_PER_LEVEL],
	lcliff_vaos[ROADS_PER_LEVEL];
// vertex point counts for each VAO
GLuint road_point_count, rcliff_point_count, lcliff_point_count;
// shader programmes
GLuint road_sp, rcliff_sp;
// uniform locations
GLint road_P_loc, road_V_loc;
GLint rcliff_P_loc, rcliff_V_loc, rcliff_w_loc, rcliff_h_loc;
// textures
GLuint road_tex;

// length of each terrain block
unsigned int num_terrain_segs = 10;

//
// pre-generate as much as possible
void init_terrain () {
	int i;
	vec3 start_left (0.0f, 0.0f, 0.0f);

	printf ("Init terrain...\n");
	srand (time (NULL));
	road_point_count = num_terrain_segs * 6;
	lcliff_point_count = num_terrain_segs * 6;
	rcliff_point_count = num_terrain_segs * 12;
	
	road_tex = create_texture_from_file (ROAD_TEX);
	
	road_sp = link_programme_from_files (ROAD_VS, ROAD_FS);
	road_P_loc = glGetUniformLocation (road_sp, "P");
	road_V_loc = glGetUniformLocation (road_sp, "V");
	
	rcliff_sp = link_programme_from_files (RCLIFF_VS, RCLIFF_FS);
	rcliff_P_loc = glGetUniformLocation (rcliff_sp, "P");
	rcliff_V_loc = glGetUniformLocation (rcliff_sp, "V");
	rcliff_h_loc = glGetUniformLocation (rcliff_sp, "h");
	rcliff_w_loc = glGetUniformLocation (rcliff_sp, "w");
	
	glGenVertexArrays (ROADS_PER_LEVEL, road_vaos);
	glGenVertexArrays (ROADS_PER_LEVEL, rcliff_vaos);
	glGenVertexArrays (ROADS_PER_LEVEL, lcliff_vaos);
	
	// gen some terrain blocks
	for (i = 0; i < ROADS_PER_LEVEL; i++) {
		gen_terrain_block (start_left, i, &start_left);
	}
}

//
// generate a random x-offset for a road segment
float rand_road_offs () {
	float o = (float)rand () / (float)RAND_MAX;
	o = o * 2.0 - 1.0; // -0.1 to + 0.1
	return o;
}

//
// creates new randomised geometry and stores in an existing vertex buffer
// object
bool gen_terrain_block (vec3 start_left, int vao_index, vec3* end_left) {
	// decide on type
	// 0 - straight
	// 1 - curve left
	// 2 - curve right
	// 3 - double curve
	// all types have random variation thrown in
	int type = 0;
	unsigned int i = 0;
	float* vps = NULL;
	float* vts = NULL;
	float* rcliff_vps = NULL;
	float* lcliff_vps = NULL;
	GLuint vp_vbo, vt_vbo, rcliff_vp_vbo, lcliff_vp_vbo;
	GLuint vps_sz, vts_sz, rcliff_vps_sz, lcliff_vps_sz;
	float curr_x, curr_y, curr_z;
	
	curr_x = start_left.v[0];
	curr_y = start_left.v[1];
	curr_z = start_left.v[2];
	
	glGenBuffers (1, &vp_vbo);
	glGenBuffers (1, &vt_vbo);
	glGenBuffers (1, &rcliff_vp_vbo);
	glGenBuffers (1, &lcliff_vp_vbo);
	
	// decide if 3-seg straight of completely random
	// i added this cos we needed more straights for speed bits
	static int four_seg_counter = 0;
	static bool straight = false;
	if (four_seg_counter % 4 == 0) {
		if ((rand () % 2) == 1) {
			straight = true;
		} else {
			straight = false;
		}
	}
	if (straight) {
		type = 0;
	} else {
		type = rand () % 3;
	}
	four_seg_counter++;
	
	printf ("block type %i gend\n", type);
	
	vps_sz = num_terrain_segs * 6 * 3 * sizeof (float);
	vps = (float*)malloc (vps_sz);
	
	rcliff_vps_sz = num_terrain_segs * 12 * 3 * sizeof (float);
	rcliff_vps = (float*)malloc (rcliff_vps_sz);
	
	lcliff_vps_sz = num_terrain_segs * 6 * 3 * sizeof (float);
	lcliff_vps = (float*)malloc (lcliff_vps_sz);
	// create 2 triangle for each seg, starting at back_left, and finishing at
	// top_left
	
/*
Road segment created in buffer like this (birds-eye view of road):

4    2,3
------
|   /|
|  / |
| /  |
|/___|
5,0  1

*/
	// TODO add curve functions to X
	for (i = 0; i < num_terrain_segs; i++) {
		// this is for adding some jitter to the road/cliffs so never perfectly
		// straight
		float x_off = 0.0f;
		float rcliff_middle_x_off = 0.5f;
		float rcliff_top_x_off = 4.0f;
		float lcliff_middle_x_off = -25.0f;
		float lcliff_middle_y = -30.0f;
		vec3 rlm, llm;
		float curve_x_offs = 0.0f;
		float prog;
		
		prog = (float)i / (float)num_terrain_segs;
		if (1 == type) {
			curve_x_offs = sinf (prog * M_PI * 0.5f) * 4.0f;
		} else if (2 == type) {
			curve_x_offs = sinf (prog * M_PI * 0.5f + M_PI) * 4.0f;
		}
		x_off = rand_road_offs () + curve_x_offs;
		
		//
		// generate the road surface
		
		// point 0
		vps[i * 18 + 0] = curr_x + -2.0f;
		vps[i * 18 + 1] = curr_y + 0.0f;
		vps[i * 18 + 2] = curr_z + -1.0f * (float)(i * 4);
		// point 1
		vps[i * 18 + 3] = curr_x + 2.0f;
		vps[i * 18 + 4] = curr_y + 0.0f;
		vps[i * 18 + 5] = curr_z + -1.0f * (float)(i * 4);
		// point 2
		vps[i * 18 + 6] = curr_x + 2.0f + x_off;
		vps[i * 18 + 7] = curr_y + 0.0f;
		vps[i * 18 + 8] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 3
		vps[i * 18 + 9] = curr_x + 2.0f + x_off;
		vps[i * 18 + 10] = curr_y + 0.0f;
		vps[i * 18 + 11] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 4
		vps[i * 18 + 12] = curr_x + -2.0f + x_off;
		vps[i * 18 + 13] = curr_y + 0.0f;
		vps[i * 18 + 14] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 5
		vps[i * 18 + 15] = curr_x + -2.0f;
		vps[i * 18 + 16] = curr_y + 0.0f;
		vps[i * 18 + 17] = curr_z + -1.0f * (float)(i * 4);
		
		// update end of road
		*end_left = vec3 (
			vps[i * 18 + 9] - 2.0f,
			vps[i * 18 + 10],
			vps[i * 18 + 11]
		);
		// let traffic know where this road goes
		llm = *end_left + vec3 (-0.75f, 0.0f, 0.0f);
		add_left_lane_marker (llm);
		rlm = *end_left + vec3 (0.75f, 0.0f, 0.0f);
		add_right_lane_marker (rlm);
		
		//
		// generate the cliffs on right side
		// road points 1 and 2 are the bottom of the cliff
		
		// point 0 (road point 2 - forward)
		rcliff_vps[i * 36 + 0] = curr_x + 2.0f + x_off;
		rcliff_vps[i * 36 + 1] = curr_y + 0.0f;
		rcliff_vps[i * 36 + 2] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 1 (road point 1 - back)
		rcliff_vps[i * 36 + 3] = curr_x + 2.0f;
		rcliff_vps[i * 36 + 4] = curr_y + 0.0f;
		rcliff_vps[i * 36 + 5] = curr_z + -1.0f * (float)(i * 4);
		// point 2 (previous point but up 1)
		rcliff_vps[i * 36 + 6] = curr_x + 2.0f + rcliff_middle_x_off;
		rcliff_vps[i * 36 + 7] = curr_y + 2.0f;
		rcliff_vps[i * 36 + 8] = curr_z + -1.0f * (float)(i * 4);
		// point 3 (same again)
		rcliff_vps[i * 36 + 9] = curr_x + 2.0f + rcliff_middle_x_off;
		rcliff_vps[i * 36 + 10] = curr_y + 2.0f;
		rcliff_vps[i * 36 + 11] = curr_z + -1.0f * (float)(i * 4);
		// point 4 (previous point but forward)
		rcliff_vps[i * 36 + 12] = curr_x + 2.0f + x_off + rcliff_middle_x_off;
		rcliff_vps[i * 36 + 13] = curr_y + 2.0f;
		rcliff_vps[i * 36 + 14] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 5 (back to the zeroth point)
		rcliff_vps[i * 36 + 15] = curr_x + 2.0f + x_off;
		rcliff_vps[i * 36 + 16] = curr_y + 0.0f;
		rcliff_vps[i * 36 + 17] = curr_z + -1.0f * (float)((i + 1) * 4);
		
		//
		// 2nd row on right-hand cliff
		
		// point 6 (road point 2 - forward)
		rcliff_vps[i * 36 + 18] = curr_x + 2.0f + x_off + rcliff_middle_x_off;
		rcliff_vps[i * 36 + 19] = curr_y + 2.0f;
		rcliff_vps[i * 36 + 20] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 7 (road point 1 - back)
		rcliff_vps[i * 36 + 21] = curr_x + 2.0f + rcliff_middle_x_off;
		rcliff_vps[i * 36 + 22] = curr_y + 2.0f;
		rcliff_vps[i * 36 + 23] = curr_z + -1.0f * (float)(i * 4);
		// point 8 (previous point but up 1)
		rcliff_vps[i * 36 + 24] = curr_x + 2.0f + rcliff_top_x_off;
		rcliff_vps[i * 36 + 25] = curr_y + 12.0f;
		rcliff_vps[i * 36 + 26] = curr_z + -1.0f * (float)(i * 4);
		// point 9 (same again)
		rcliff_vps[i * 36 + 27] = curr_x + 2.0f + rcliff_top_x_off;
		rcliff_vps[i * 36 + 28] = curr_y + 12.0f;
		rcliff_vps[i * 36 + 29] = curr_z + -1.0f * (float)(i * 4);
		// point 10 (previous point but forward)
		rcliff_vps[i * 36 + 30] = curr_x + 2.0f + x_off + rcliff_top_x_off;
		rcliff_vps[i * 36 + 31] = curr_y + 12.0f;
		rcliff_vps[i * 36 + 32] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 11 (back to the zeroth point)
		rcliff_vps[i * 36 + 33] = curr_x + 2.0f + x_off + rcliff_middle_x_off;
		rcliff_vps[i * 36 + 34] = curr_y + 2.0f;
		rcliff_vps[i * 36 + 35] = curr_z + -1.0f * (float)((i + 1) * 4);
		
		//
		// left-hand side down cliffs
		//
		// point 0 (road point 2 - forward)
		lcliff_vps[i * 18 + 0] = curr_x - 2.0f + x_off + lcliff_middle_x_off;
		lcliff_vps[i * 18 + 1] = curr_y + lcliff_middle_y;
		lcliff_vps[i * 18 + 2] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 1 (road point 1 - back)
		lcliff_vps[i * 18 + 3] = curr_x - 2.0f + lcliff_middle_x_off;
		lcliff_vps[i * 18 + 4] = curr_y + lcliff_middle_y;
		lcliff_vps[i * 18 + 5] = curr_z + -1.0f * (float)(i * 4);
		// point 2 (previous point but up 1)
		lcliff_vps[i * 18 + 6] = curr_x - 2.0f;
		lcliff_vps[i * 18 + 7] = curr_y;
		lcliff_vps[i * 18 + 8] = curr_z + -1.0f * (float)(i * 4);
		// point 3 (same again)
		lcliff_vps[i * 18 + 9] = curr_x - 2.0f;
		lcliff_vps[i * 18 + 10] = curr_y;
		lcliff_vps[i * 18 + 11] = curr_z + -1.0f * (float)(i * 4);
		// point 4 (previous point but forward)
		lcliff_vps[i * 18 + 12] = curr_x - 2.0f + x_off;
		lcliff_vps[i * 18 + 13] = curr_y;
		lcliff_vps[i * 18 + 14] = curr_z + -1.0f * (float)((i + 1) * 4);
		// point 5 (back to the zeroth point)
		lcliff_vps[i * 18 + 15] = curr_x - 2.0f + x_off + lcliff_middle_x_off;
		lcliff_vps[i * 18 + 16] = curr_y + lcliff_middle_y;
		lcliff_vps[i * 18 + 17] = curr_z + -1.0f * (float)((i + 1) * 4);
		
		curr_x = curr_x + x_off;
		// TODO z as well instead of i *
	}
	glBindBuffer (GL_ARRAY_BUFFER, vp_vbo);
	glBufferData (GL_ARRAY_BUFFER, vps_sz, vps, GL_STATIC_DRAW);
	free (vps);
	
	glBindBuffer (GL_ARRAY_BUFFER, rcliff_vp_vbo);
	glBufferData (GL_ARRAY_BUFFER, rcliff_vps_sz, rcliff_vps, GL_STATIC_DRAW);
	free (rcliff_vps);
	
	glBindBuffer (GL_ARRAY_BUFFER, lcliff_vp_vbo);
	glBufferData (GL_ARRAY_BUFFER, lcliff_vps_sz, lcliff_vps, GL_STATIC_DRAW);
	free (lcliff_vps);
	
	vts_sz = num_terrain_segs * 6 * 2 * sizeof (float);
	vts = (float*)malloc (vts_sz);
	for (i = 0; i < num_terrain_segs; i++) {
		// point 0
		vts[i * 12 + 0] = 0.0f;
		vts[i * 12 + 1] = 0.0f;
		// point 1
		vts[i * 12 + 2] = 1.0f;
		vts[i * 12 + 3] = 0.0f;
		// point 2
		vts[i * 12 + 4] = 1.0f;
		vts[i * 12 + 5] = 1.0f;
		// point 3
		vts[i * 12 + 6] = 1.0f;
		vts[i * 12 + 7] = 1.0f;
		// point 4
		vts[i * 12 + 8] = 0.0f;
		vts[i * 12 + 9] = 1.0f;
		// point 5
		vts[i * 12 + 10] = 0.0f;
		vts[i * 12 + 11] = 0.0f;
	}
	glBindBuffer (GL_ARRAY_BUFFER, vt_vbo);
	glBufferData (GL_ARRAY_BUFFER, vts_sz, vts, GL_STATIC_DRAW);
	free (vts);
	
	glBindVertexArray (road_vaos[vao_index]);
	glEnableVertexAttribArray (0);
	glBindBuffer (GL_ARRAY_BUFFER, vp_vbo);
	glVertexAttribPointer (0, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	glEnableVertexAttribArray (1);
	glBindBuffer (GL_ARRAY_BUFFER, vt_vbo);
	glVertexAttribPointer (1, 2, GL_FLOAT, GL_FALSE, 0, NULL);
	
	glBindVertexArray (rcliff_vaos[vao_index]);
	glEnableVertexAttribArray (0);
	glBindBuffer (GL_ARRAY_BUFFER, rcliff_vp_vbo);
	glVertexAttribPointer (0, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	
	glBindVertexArray (lcliff_vaos[vao_index]);
	glEnableVertexAttribArray (0);
	glBindBuffer (GL_ARRAY_BUFFER, lcliff_vp_vbo);
	glVertexAttribPointer (0, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	
	return true;
}

void draw_terrain () {
	int i;
	vec3 player_pos;

	player_pos = get_player_pos ();
	glActiveTexture (GL_TEXTURE0);
	glBindTexture (GL_TEXTURE_2D, road_tex);
	
	for (i = 0; i < ROADS_PER_LEVEL; i++) {
		float tz;
		
		// don't draw behind camera
		tz = (float)i * -4.0f * 10.0f;
		if (is_forward_cam) {
			if (tz > player_pos.v[2] + 40.0f) {
				continue;
			}
			if (tz < player_pos.v[2] - 120.0f) {
				continue;
			}
		} else {
			if (tz < player_pos.v[2] - 40.0f) {
				continue;
			}
			if (tz > player_pos.v[2] + 80.0f) {
				continue;
			}
		}
	
	
		glUseProgram (road_sp);
		if (cam_V_dirty) {
			glUniformMatrix4fv (road_V_loc, 1, GL_FALSE, V.m);
			uniforms++;
		}
		if (cam_P_dirty) {
			glUniformMatrix4fv (road_P_loc, 1, GL_FALSE, P.m);
			uniforms++;
		}
	
		glBindVertexArray (road_vaos[i]);
		glDrawArrays (GL_TRIANGLES, 0, road_point_count);
		draws++;
		verts += road_point_count;
	
		glUseProgram (rcliff_sp);
		if (cam_V_dirty) {
			glUniformMatrix4fv (rcliff_V_loc, 1, GL_FALSE, V.m);
			uniforms++;
		}
		if (cam_P_dirty) {
			glUniformMatrix4fv (rcliff_P_loc, 1, GL_FALSE, P.m);
			glUniform1f (rcliff_h_loc, (float)gl_height);
			glUniform1f (rcliff_w_loc, (float)gl_width);
			uniforms += 3;
		}
	
		glBindVertexArray (rcliff_vaos[i]);
		glDrawArrays (GL_TRIANGLES, 0, rcliff_point_count);
		draws++;
		verts += rcliff_point_count;
		
		glBindVertexArray (lcliff_vaos[i]);
		glDrawArrays (GL_TRIANGLES, 0, lcliff_point_count);
		draws++;
		verts += lcliff_point_count;
	}
}

